### ISSUE TITLE
Segfault on upgrade from 0.5.1 to 0.5.4

### ISSUE THREAD

**Issue by grobinson-blockchain** [hashicorp/nomad#2295]
### Nomad version
Nomad v0.5.4

### Issue
Upgrading Nomad from 0.5.1. to 0.5.4 causes a segfault:

    Feb 09 12:14:03 backend-generic-xxx nomad[9570]: panic: runtime error: invalid memory address or nil pointer dereference
    Feb 09 12:14:03 backend-generic-xxx nomad[9570]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5ebe54]
    Feb 09 12:14:03 backend-generic-xxx nomad[9570]: goroutine 121 [running]:
    Feb 09 12:14:03 backend-generic-xxx nomad[9570]: panic(0x1000940, 0xc420010090)
    Feb 09 12:14:03 backend-generic-xxx nomad[9570]: /opt/go/src/runtime/panic.go:500 +0x1a1
    Feb 09 12:14:03 backend-generic-xxx nomad[9570]: github.com/hashicorp/nomad/client/driver.(*CreatedResources).Merge(0x0, 0xc4200661d0)
    Feb 09 12:14:03 backend-generic-xxx nomad[9570]: /opt/gopath/src/github.com/hashicorp/nomad/client/driver/driver.go:131 +0xe4
    Feb 09 12:14:03 backend-generic-xxx nomad[9570]: github.com/hashicorp/nomad/client.(*TaskRunner).startTask(0xc42025c000, 0xc4204274a0, 0x0)
    Feb 09 12:14:03 backend-generic-xxx nomad[9570]: /opt/gopath/src/github.com/hashicorp/nomad/client/task_runner.go:1169 +0x2e3
    Feb 09 12:14:03 backend-generic-xxx nomad[9570]: github.com/hashicorp/nomad/client.(*TaskRunner).run(0xc42025c000)
    Feb 09 12:14:03 backend-generic-xxx nomad[9570]: /opt/gopath/src/github.com/hashicorp/nomad/client/task_runner.go:902 +0x38a
    Feb 09 12:14:03 backend-generic-xxx nomad[9570]: github.com/hashicorp/nomad/client.(*TaskRunner).Run(0xc42025c000)
    Feb 09 12:14:03 backend-generic-xxx nomad[9570]: /opt/gopath/src/github.com/hashicorp/nomad/client/task_runner.go:444 +0x6a1
    Feb 09 12:14:03 backend-generic-xxx nomad[9570]: created by github.com/hashicorp/nomad/client.(*AllocRunner).RestoreState
    Feb 09 12:14:03 backend-generic-xxx nomad[9570]: /opt/gopath/src/github.com/hashicorp/nomad/client/alloc_runner.go:190 +0x891

**Comment by schmichael** [comment]
I'm sorry to hear that @grobinson-blockchain! I'm having trouble reproducing it locally. What types of jobs were running on this node? Were there stopped/dead/completed jobs/tasks on the node? Did you try upgrading more than one node and run into this every time?

**Comment by rileyje** [comment]
We had the same issue upgrading from 0.5.2 to 0.5.4. The masters did not have any issues during upgrade. Not all clients were affected, about 13/32 have the problem. At least some of the nodes having the issue had running Docker allocations. They likely all had stopped/dead/completed tasks.

```
Feb 09 18:35:59 node-2 systemd[1]: Started nomad.
Feb 09 18:35:59 node-2 systemd[1]: Starting nomad...
Feb 09 18:35:59 node-2 nomad[16463]: Loaded configuration from /opt/nomad/nomad.d/client.hcl, /opt/nomad/nomad.d/nomad.hcl
Feb 09 18:35:59 node-2 nomad[16463]: ==> Starting Nomad agent...
Feb 09 18:36:03 node-2 nomad[16463]: panic: runtime error: invalid memory address or nil pointer dereference
Feb 09 18:36:03 node-2 nomad[16463]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5ebe54]
Feb 09 18:36:03 node-2 nomad[16463]: goroutine 92 [running]:
Feb 09 18:36:03 node-2 nomad[16463]: panic(0x1000940, 0xc420010050)
Feb 09 18:36:03 node-2 nomad[16463]: /opt/go/src/runtime/panic.go:500 +0x1a1
Feb 09 18:36:03 node-2 nomad[16463]: github.com/hashicorp/nomad/client/driver.(*CreatedResources).Merge(0x0, 0xc4203dc108)
Feb 09 18:36:03 node-2 nomad[16463]: /opt/gopath/src/github.com/hashicorp/nomad/client/driver/driver.go:131 +0xe4
Feb 09 18:36:03 node-2 nomad[16463]: github.com/hashicorp/nomad/client.(*TaskRunner).startTask(0xc4204958c0, 0xc420335980, 0x0)
Feb 09 18:36:03 node-2 nomad[16463]: /opt/gopath/src/github.com/hashicorp/nomad/client/task_runner.go:1169 +0x2e3
Feb 09 18:36:03 node-2 nomad[16463]: github.com/hashicorp/nomad/client.(*TaskRunner).run(0xc4204958c0)
Feb 09 18:36:03 node-2 nomad[16463]: /opt/gopath/src/github.com/hashicorp/nomad/client/task_runner.go:902 +0x38a
Feb 09 18:36:03 node-2 nomad[16463]: github.com/hashicorp/nomad/client.(*TaskRunner).Run(0xc4204958c0)
Feb 09 18:36:03 node-2 nomad[16463]: /opt/gopath/src/github.com/hashicorp/nomad/client/task_runner.go:444 +0x6a1
Feb 09 18:36:03 node-2 nomad[16463]: created by github.com/hashicorp/nomad/client.(*AllocRunner).RestoreState
Feb 09 18:36:03 node-2 nomad[16463]: /opt/gopath/src/github.com/hashicorp/nomad/client/alloc_runner.go:190 +0x891
Feb 09 18:36:03 node-2 systemd[1]: nomad.service: main process exited, code=exited, status=2/INVALIDARGUMENT
Feb 09 18:36:03 node-2 systemd[1]: Unit nomad.service entered failed state.
Feb 09 18:36:03 node-2 systemd[1]: nomad.service failed.
Feb 09 18:36:04 node-2 systemd[1]: nomad.service holdoff time over, scheduling restart.
```

**Comment by rileyje** [comment]
Found another client with more information in the journald log. Happened to be running with DEBUG logging. We can leave it broken if you have questions about the allocations listed in the output.

```
Feb 09 18:55:22 node-15 nomad[27076]: ==> Nomad agent configuration:
Feb 09 18:55:22 node-15 nomad[27076]: Atlas: <disabled>
Feb 09 18:55:22 node-15 nomad[27076]: Client: true
Feb 09 18:55:22 node-15 nomad[27076]: Log Level: DEBUG
Feb 09 18:55:22 node-15 nomad[27076]: Region: local (DC: ch4)
Feb 09 18:55:22 node-15 nomad[27076]: Server: false
Feb 09 18:55:22 node-15 nomad[27076]: Version: 0.5.4
Feb 09 18:55:22 node-15 nomad[27076]: ==> Nomad agent started! Log data will stream in below:
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:18.103346 [INFO] client: using state directory /opt/nomad/lib/client
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:18.103408 [INFO] client: using alloc directory /opt/nomad/lib/alloc
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:18.103491 [DEBUG] client: built-in fingerprints: [arch cgroup consul cpu host memory network nomad signal storage vault env_aws env_gce]
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:18.103667 [INFO] fingerprint.cgroups: cgroups are available
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:18.103810 [DEBUG] client: fingerprinting cgroup every 15s
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:18.104940 [INFO] fingerprint.consul: consul agent is available
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:18.105028 [DEBUG] client: fingerprinting consul every 15s
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:18.106807 [DEBUG] fingerprint.cpu: frequency: 2933 MHz
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:18.106812 [DEBUG] fingerprint.cpu: core count: 12
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:18.121935 [DEBUG] fingerprint.network: Detected interface bond0 with IP 10.104.105.64 during fingerprinting
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:18.122764 [DEBUG] fingerprint.network: link speed for bond0 set to 10000
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:18.124021 [DEBUG] client: fingerprinting vault every 15s
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:20.124086 [DEBUG] fingerprint.env_gce: Could not read value for attribute "machine-type"
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:20.124093 [DEBUG] fingerprint.env_gce: Error querying GCE Metadata URL, skipping
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.124255 [DEBUG] fingerprint.env_aws: Error querying AWS Metadata URL, skipping
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.124278 [DEBUG] client: applied fingerprints [arch cgroup consul cpu host memory network nomad signal storage]
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.124310 [DEBUG] driver.docker: using client connection initialized from environment
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.125661 [DEBUG] driver.exec: exec driver is enabled
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.125742 [DEBUG] client: available drivers [docker exec]
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.126271 [DEBUG] client: fingerprinting exec every 15s
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.126288 [DEBUG] client: fingerprinting docker every 15s
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.126300 [DEBUG] client: fingerprinting rkt every 15s
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.127195 [DEBUG] client.consul: bootstrap contacting following Consul DCs: ["bc-ch4-103a"]
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.127523 [DEBUG] client: alloc "08ddbfc9-984b-ccb3-025e-7070f4678a7c" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.127956 [DEBUG] client: alloc "232c12b1-f0ce-4853-6320-95256923ba78" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.128355 [DEBUG] client: alloc "2aebc457-2018-8846-d7ce-470425a98e38" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.128756 [DEBUG] client: alloc "46ba2895-48f3-0d54-e0ad-245401c5622f" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.129076 [DEBUG] client: alloc "53eb5f93-908e-5a8d-51a6-da38fca86953" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.129119 [INFO] client.consul: discovered following Servers: 10.104.105.50:4647,10.104.105.53:4647,10.104.105.56:4647
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.129439 [DEBUG] client: alloc "58626e56-f57c-7080-c359-ad3edeba4b13" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.129616 [DEBUG] client: migrating state snapshot for alloc "5a63f4fb-d7d3-3472-9281-e22a9d4a68ff"
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.129796 [DEBUG] client: alloc "5a63f4fb-d7d3-3472-9281-e22a9d4a68ff" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.129970 [DEBUG] client: migrating state snapshot for alloc "5e43463d-6f5c-00cc-9a71-5a77acf2e725"
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.130174 [DEBUG] client: alloc "5e43463d-6f5c-00cc-9a71-5a77acf2e725" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.130513 [DEBUG] client: alloc "7e7aa7f5-5af5-6cda-4107-ba3dd48bf330" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 systemd[1]: nomad.service: main process exited, code=exited, status=2/INVALIDARGUMENT
Feb 09 18:55:22 node-15 systemd[1]: Unit nomad.service entered failed state.
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.130674 [DEBUG] client: migrating state snapshot for alloc "8553f850-27f7-4873-cea2-abea447cf33e"
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.130833 [DEBUG] client: alloc "8553f850-27f7-4873-cea2-abea447cf33e" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.131175 [DEBUG] client: alloc "87e5c7bd-ff0a-693d-2b42-4a19c43f3718" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.131513 [DEBUG] client: alloc "88b62845-0348-b574-ad3c-b2adddce1056" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.131675 [DEBUG] client: migrating state snapshot for alloc "99b131dc-3f8c-1d1e-426c-d6c3e8cef3a8"
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.131864 [DEBUG] client: alloc "99b131dc-3f8c-1d1e-426c-d6c3e8cef3a8" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.132172 [DEBUG] client: alloc "a3b01169-60e6-4118-10cb-27c661111e76" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.132546 [DEBUG] client: alloc "ae79b088-d26d-512c-668b-4bc804ef675a" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.132886 [DEBUG] client: alloc "bce9fa6a-ac0c-094f-36e6-f18a32f9e1b5" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.133366 [DEBUG] client: starting task context for 'compute-node-test' (alloc 'c0e8ae9e-beae-a70d-ae5a-cbabc7c0e65a')
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.133945 [DEBUG] client: migrating state snapshot for alloc "e8efd872-36b3-65f1-c97e-2bbdf66b4fc6"
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.133953 [DEBUG] client: starting task runners for alloc 'c0e8ae9e-beae-a70d-ae5a-cbabc7c0e65a'
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.134055 [DEBUG] client: alloc "d9558d71-5561-a083-2603-0ddf6dd6a42e" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.134111 [DEBUG] client: alloc "e8efd872-36b3-65f1-c97e-2bbdf66b4fc6" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.134370 [DEBUG] client: alloc "f59d6df4-a09a-e84e-bc40-63c3c2295cc8" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.134665 [INFO] client: Node ID "c8dcba5d-4ae0-e800-93b0-0615e32661a3"
Feb 09 18:55:22 node-15 nomad[27076]: 2017/02/09 18:55:22.134675 [DEBUG] client: alloc "fc68fb5e-b82e-081b-16f6-e34499b6bf3d" in terminal status, waiting for destroy
Feb 09 18:55:22 node-15 nomad[27076]: panic: runtime error: invalid memory address or nil pointer dereference
Feb 09 18:55:22 node-15 nomad[27076]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5ebe54]
Feb 09 18:55:22 node-15 nomad[27076]: goroutine 119 [running]:
Feb 09 18:55:22 node-15 nomad[27076]: panic(0x1000940, 0xc420010050)
Feb 09 18:55:22 node-15 nomad[27076]: /opt/go/src/runtime/panic.go:500 +0x1a1
Feb 09 18:55:22 node-15 nomad[27076]: github.com/hashicorp/nomad/client/driver.(*CreatedResources).Merge(0x0, 0xc420136758)
Feb 09 18:55:22 node-15 nomad[27076]: /opt/gopath/src/github.com/hashicorp/nomad/client/driver/driver.go:131 +0xe4
Feb 09 18:55:22 node-15 nomad[27076]: github.com/hashicorp/nomad/client.(*TaskRunner).startTask(0xc4203f6f20, 0xc4200ffa40, 0x0)
Feb 09 18:55:22 node-15 nomad[27076]: /opt/gopath/src/github.com/hashicorp/nomad/client/task_runner.go:1169 +0x2e3
Feb 09 18:55:22 node-15 nomad[27076]: github.com/hashicorp/nomad/client.(*TaskRunner).run(0xc4203f6f20)
Feb 09 18:55:22 node-15 nomad[27076]: /opt/gopath/src/github.com/hashicorp/nomad/client/task_runner.go:902 +0x38a
Feb 09 18:55:22 node-15 nomad[27076]: github.com/hashicorp/nomad/client.(*TaskRunner).Run(0xc4203f6f20)
Feb 09 18:55:22 node-15 nomad[27076]: /opt/gopath/src/github.com/hashicorp/nomad/client/task_runner.go:444 +0x6a1
Feb 09 18:55:22 node-15 nomad[27076]: created by github.com/hashicorp/nomad/client.(*AllocRunner).RestoreState
Feb 09 18:55:22 node-15 nomad[27076]: /opt/gopath/src/github.com/hashicorp/nomad/client/alloc_runner.go:190 +0x891
Feb 09 18:55:22 node-15 systemd[1]: nomad.service failed.
Feb 09 18:55:22 node-15 systemd[1]: nomad.service holdoff time over, scheduling restart.
```

**Comment by schmichael** [comment]
@rileyje Thanks for the extra info! Not every node being affected lines up with my suspicion as to the root cause. #2301 fixes it and will be in 0.5.5.

Sorry for the hassle.
